File: springboot/src/main/java/com/edufelip/meer/domain/auth/SignupUseCase.java
package com.edufelip.meer.domain.auth;

import com.edufelip.meer.core.auth.AuthUser;
import com.edufelip.meer.domain.repo.AuthUserRepository;
import com.edufelip.meer.security.token.TokenProvider;
import org.springframework.security.crypto.password.PasswordEncoder;

public class SignupUseCase {
    private final AuthUserRepository authUserRepository;
    private final PasswordEncoder passwordEncoder;
    private final TokenProvider tokenProvider;

    public SignupUseCase(AuthUserRepository authUserRepository, PasswordEncoder passwordEncoder, TokenProvider tokenProvider) {
        this.authUserRepository = authUserRepository;
        this.passwordEncoder = passwordEncoder;
        this.tokenProvider = tokenProvider;
    }

    public AuthResult execute(String name, String email, String password) {
        if (authUserRepository.findByEmail(email) != null) throw new EmailAlreadyRegisteredException();

        String hashed = passwordEncoder.encode(password);
        AuthUser user = authUserRepository.save(new AuthUser(email, name, null, hashed));

        String access = tokenProvider.generateAccessToken(user);
        String refresh = tokenProvider.generateRefreshToken(user);

        AuthenticatedUser authUser = new AuthenticatedUser(user.getId(), user.getDisplayName(), user.getEmail());
        return new AuthResult(access, refresh, authUser);
    }
}

File: springboot/src/main/kotlin/com/edufelip/meer/domain/auth/SignupUseCase.kt
package com.edufelip.meer.domain.auth

import com.edufelip.meer.core.auth.AuthUser
import com.edufelip.meer.domain.AuthUserRepository
import com.edufelip.meer.security.token.TokenProvider
import org.springframework.security.crypto.password.PasswordEncoder

class SignupUseCase(
    private val authUserRepository: AuthUserRepository,
    private val passwordEncoder: PasswordEncoder,
    private val tokenProvider: TokenProvider
) {
    fun execute(name: String, email: String, password: String): AuthResult {
        if (authUserRepository.findByEmail(email) != null) throw EmailAlreadyRegisteredException()

        val hashedPassword: String = passwordEncoder.encode(password).toString()

        val user = AuthUser(
            email = email,
            displayName = name,
            passwordHash = hashedPassword
        )

        val saved = authUserRepository.save(user)
        val accessToken = tokenProvider.generateAccessToken(saved)
        val refreshToken = tokenProvider.generateRefreshToken(saved)

        return AuthResult(
            token = accessToken,
            refreshToken = refreshToken,
            user = AuthenticatedUser(id = saved.id, name = saved.displayName, email = saved.email)
        )
    }
}

class EmailAlreadyRegisteredException : RuntimeException("Email already registered")
